src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -225,19 +268,259 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+                    () => this.ResponseStream,
+                    throwOnFailure: true,
+                    out responseVersion);
+
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+
+                ResponseInfo responseInfo = this.CreateResponseInfo(entityDescriptor);
+                HttpWebResponseMessage responseMessageWrapper = new HttpWebResponseMessage(
+                    headers,
+                    this.batchResponseMessage.StatusCode,
+                    () => this.ResponseStream);
+
+                ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                entry = ODataReaderEntityMaterializer.ParseSingleEntityPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+
+                this.headers = headers;
+                this.topLevelEntry = entry;
+            }
+            catch (DataServiceRequestException)
+            {
+                throw;
+            }
+            catch (InvalidOperationException ex)
+            {
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+                int statusCode = this.batchResponseMessage == null ? (int)HttpStatusCode.InternalServerError : (int)this.batchResponseMessage.StatusCode;
+                DataServiceResponse dataServiceResponse = new DataServiceResponse(headers, statusCode, Enumerable.Empty<OperationResponse>(), this.IsBatchRequest);
+                throw new DataServiceRequestException(Strings.DataServiceException_GeneralError, ex, dataServiceResponse);
+            }
+
+            HandleDeepInsertResponseInternal(entry, true, entityDescriptor, null);
+
+            return response;
+        }
+
+        private void HandleDeepInsertResponseInternal(MaterializerEntry entry, bool isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse)
+        {
+            OperationResponse response = this.CreateOperationResponse(entry, isTopLevelDescriptor, descriptor, parentOperationResponse);
+
+            List<Descriptor> relatedDescriptors = this.BulkUpdateGraph.GetRelatedDescriptors(descriptor);
+
+            foreach (IMaterializerState nestedItem in entry?.NestedItems)
+            {
+                if (nestedItem is MaterializerNestedEntry materializerNestedEntry)
+                {
+                    foreach (IMaterializerState nestedEntries in materializerNestedEntry.NestedItems)
+                    {
+                        if (nestedEntries is MaterializerFeed feed)
+                        {
+                            for (int i = 0; i < feed.Items.Count; i++)
+                            {
+                                // Ensures the count of the descriptors matches the feed.Items count.
+                                if (relatedDescriptors.Count < i + 1)

```

Comment:
Question: In what scenarios would the number of items in the feed be different from the number of items in the `relatedDescriptors` collection? Trying to understand whether we should silent break and move or raise an exception...

Reply:
If the descriptors have been detached